File: src/io/coconut/ferrum/utils/Checker.scala
package io.coconut.ferrum.utils

import com.google.gson.{JsonArray, JsonObject}
import io.coconut.ferrum.Launcher
import io.coconut.ferrum.types.*

object Checker {
  def whichType(version: String): String = {
    val versionType = if (version.startsWith("b") || version.contains("beta")) {
      VersionType.BETA
    } else if (version.startsWith("a") || version.contains("alpha")) {
      VersionType.ALPHA
    } else if (version.contains("w") || version.contains("-pre") || version.contains("-rc")) {
      VersionType.SNAPSHOT
    } else {
      VersionType.RELEASE
    }

    versionType.secondName
  }

  def isByRules(rules: JsonArray): Boolean = {
    val os = System.getProperty("os.name").toLowerCase

    import scala.util.boundary
    import scala.util.boundary.break

    boundary:
      for (i <- 0 until rules.size()) {
        val rule = rules.get(i).getAsJsonObject
        val action = rule.get("action").getAsString

        if (rule.has("os")) {
          val osRule = rule.getAsJsonObject("os")
          val osName = if (osRule.has("name")) osRule.get("name").getAsString else ""

          var matches = false
          if (osName.equals("windows") && os.contains("win")) matches = true
          if (osName.equals("linux") && os.contains("nux")) matches = true
          if (osName.equals("osx") && os.contains("mac")) matches = true

          if (action.equals("allow") && !matches) break(false)
          if (action.equals("disallow") && matches) break(false)
        }
      }

      true
  }
}

File: src/main/java/ferrum/utils/Checker.java
package ferrum.utils;

import ferrum.Version;
import org.json.JSONArray;
import org.json.JSONObject;

public class Checker {
    public static String whichType(String version) {
        VersionType type;

        if (version.startsWith("b") || version.contains("beta")) {
            type = Version.BETA;
        } else if (version.startsWith("a") || version.contains("alpha")) {
            type = VersionType.ALPHA;
        } else if (version.contains("w") || version.contains("-pre") || version.contains("-rc")) {
            type = VersionType.SNAPSHOT;
        } else {
            type = VersionType.RELEASE;
        }

        return type.getFriendlyName();
    }

    public static boolean isByRules(JSONArray rules) {
        String os = System.getProperty("os.name").toLowerCase();

        for (int i = 0; i < rules.length(); i++) {
            JSONObject rule = rules.getJSONObject(i);
            String action = rule.getString("action");

            if (rule.has("os")) {
                JSONObject osRule = rule.getJSONObject("os");
                String osName = osRule.optString("name", "");

                boolean matches = false;
                if (osName.equals("windows") && os.contains("win")) matches = true;
                if (osName.equals("linux") && os.contains("nux")) matches = true;
                if (osName.equals("osx") && os.contains("mac")) matches = true;

                if (action.equals("allow") && !matches) return false;
                if (action.equals("disallow") && matches) return false;
            }
        }

        return true;
    }
}
